File: src/main/java/cad97/spawnercraft/init/SpawnerCraftItems.java
package cad97.spawnercraft.init;

import cad97.spawnercraft.items.ItemMobAgglomeration;
import cad97.spawnercraft.items.ItemMobEssence;
import cad97.spawnercraft.items.ItemMobRod;
import cad97.spawnercraft.items.ItemMobSpirit;
import cad97.spawnercraft.utility.LogHelper;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.client.renderer.color.ItemColors;
import net.minecraft.entity.EntityList;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemMonsterPlacer;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.fml.common.registry.GameRegistry;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

public class SpawnerCraftItems {

    public static final ItemMobEssence MOB_ESSENCE = new ItemMobEssence();
    @SuppressWarnings("WeakerAccess")
    public static final ItemMobAgglomeration MOB_AGGLOMERATION = new ItemMobAgglomeration();
    @SuppressWarnings("WeakerAccess")
    public static final ItemMobSpirit MOB_SPIRIT = new ItemMobSpirit();
    public static final ItemMobRod MOB_ROD = new ItemMobRod();

    public static void registerItems() {
        GameRegistry.register(MOB_ESSENCE);
        GameRegistry.register(MOB_AGGLOMERATION);
        GameRegistry.register(MOB_SPIRIT);
        GameRegistry.register(MOB_ROD);
        GameRegistry.register(new ItemBlock(SpawnerCraftBlocks.MOB_CAGE), SpawnerCraftBlocks.MOB_CAGE.getRegistryName());
        LogHelper.logInfo("Items initialized.");
    }

    @SideOnly(Side.CLIENT)
    @SuppressWarnings("ConstantConditions")
    public static void registerModels() {
        ModelLoader.setCustomModelResourceLocation(MOB_ESSENCE, 0,
                new ModelResourceLocation(MOB_ESSENCE.getRegistryName(), "inventory"));
        ModelLoader.setCustomModelResourceLocation(MOB_AGGLOMERATION, 0,
                new ModelResourceLocation(MOB_AGGLOMERATION.getRegistryName(), "inventory"));
        ModelLoader.setCustomModelResourceLocation(MOB_SPIRIT, 0,
                new ModelResourceLocation(MOB_SPIRIT.getRegistryName(), "inventory"));
        ModelLoader.setCustomModelResourceLocation(MOB_ROD, 0,
                new ModelResourceLocation(MOB_ROD.getRegistryName(), "inventory"));
        LogHelper.logInfo("Item models initialized.");
    }

    @SideOnly(Side.CLIENT)
    public static void registerColors(ItemColors itemColors) {
        itemColors.registerItemColorHandler((stack, tintIndex) -> {
            EntityList.EntityEggInfo eggInfo = EntityList.ENTITY_EGGS.get(ItemMonsterPlacer.getNamedIdFrom(stack));
            return eggInfo == null ? -1 : (tintIndex == 0 ? eggInfo.primaryColor : eggInfo.secondaryColor);
        }, MOB_ESSENCE, MOB_AGGLOMERATION, MOB_SPIRIT);
        LogHelper.logInfo("Item colors initialized.");
    }
}

File: src/main/kotlin/cad97/spawnercraftkt/init/SpawnerCraftItems.kt
package cad97.spawnercraftkt.init

import cad97.spawnercraftkt.SpawnerCraft
import cad97.spawnercraftkt.item.ItemMobSoul
import cad97.spawnercraftkt.item.ItemMobSpirit
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.client.renderer.color.ItemColors
import net.minecraft.item.Item
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.registry.GameRegistry
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

@Mod.EventBusSubscriber(modid = SpawnerCraft.modid)
object SpawnerCraftItems {
    @JvmStatic
    @GameRegistry.ObjectHolder("${SpawnerCraft.modid}:${ItemMobSoul.essenceId}")
    lateinit var mob_essence: Item private set

    @JvmStatic
    @GameRegistry.ObjectHolder("${SpawnerCraft.modid}:${ItemMobSoul.agglomerationId}")
    lateinit var mob_agglomeration: Item private set

    @JvmStatic
    @GameRegistry.ObjectHolder("${SpawnerCraft.modid}:${ItemMobSpirit.id}")
    lateinit var mob_spirit: Item private set

    @JvmStatic
    @SubscribeEvent
    fun registerItems(event: RegistryEvent.Register<Item>) {
        for (id in setOf(ItemMobSoul.essenceId, ItemMobSoul.agglomerationId)) {
            event.registry.register(
                    ItemMobSoul()
                            .setRegistryName(SpawnerCraft.modid, id)
                            .setUnlocalizedName("${SpawnerCraft.modid}.$id")
            )
        }
        event.registry.register(ItemMobSpirit())
        SpawnerCraft.logger.info("Items registered.")
    }

    @JvmStatic
    @SubscribeEvent
    fun registerModels(event: ModelRegistryEvent) {
        for (item in setOf(mob_essence, mob_agglomeration, mob_spirit)) {
            ModelLoader.setCustomModelResourceLocation(
                    item, 0,
                    ModelResourceLocation(item.registryName.toString())
            )
        }
        SpawnerCraft.logger.info("Item models registered.")
    }

    @SideOnly(Side.CLIENT)
    fun registerColors(itemColors: ItemColors) {
        itemColors.registerItemColorHandler(ItemMobSoul.colorHandler, mob_essence, mob_agglomeration, mob_spirit)
        SpawnerCraft.logger.info("Item colors registered.")
    }
}